extensions/cornerstone-dynamic-volume/src/panels/PanelGenerateImage.tsx

```diff
@@ -0,0 +1,414 @@
+      setComputedDisplaySet(obj);
+    } else {
+      // update vtkOpenGLTexture and imageData of computed volume
+      const { imageData, vtkOpenGLTexture } = computedVolume;
+      const numSlices = imageData.getDimensions()[2];
+      const slicesToUpdate = [...Array(numSlices).keys()];
+      slicesToUpdate.forEach(i => {
+        vtkOpenGLTexture.setUpdatedFrame(i);
+      });
+      imageData.modified();
+
+      // Check if viewport is currently displaying the computed volume, if so,
+      // call render on the viewports to update the image, if not, call
+      // renderGeneratedImage
+      if (!cache.getVolume(dynamicVolumeId)) {
+        for (const viewport of viewports) {
+          const renderingEngine = cornerstoneViewportService.getRenderingEngine();
+          const viewportForRendering = renderingEngine.getViewport(
+            viewport.viewportId
+          );
+          viewportForRendering.render();
+        }
+      } else {
+        renderGeneratedImage(computedDisplaySet);
+      }
+    }
+  }
+
+  // if computedDisplaySet is defined, render when the data changes
+  useEffect(() => {
+    if (computedDisplaySet) {
+      renderGeneratedImage(computedDisplaySet);
+    }
+  }, [computedDisplaySet]);
+
+  function renderGeneratedImage(displaySet) {
+    commandsManager.runCommand('setDisplaySetToViewport', {
+      displaySet,
+    });
+  }
+
+  function returnTo4D() {
+    const displaySets = displaySetService.getDisplaySetCache();
+    for (const [key, value] of displaySets) {
+      if (value.volumeLoaderSchema === volumeLoaderScheme) {
+        const updatedViewports = hangingProtocolService.getViewportsRequireUpdate(
+          0,
+          key
+        );
+        viewportGridService.setDisplaySetsForViewports(updatedViewports);
+      }
+    }
+  }
+
+  const handlePlay = () => {
+    const viewportInfo = cornerstoneViewportService.viewportsById.get(
+      viewports[activeViewportIndex].viewportId
+    );
+    const { element } = viewportInfo;
+    const cine = cines[activeViewportIndex];
+    if (!cine.isPlaying) {
+      cineService.playClip(element, { framesPerSecond: frameRate });
+      cine.isPlaying = true;
+    }
+  };
+
+  const handleStop = () => {
+    const viewportInfo = cornerstoneViewportService.viewportsById.get(
+      viewports[activeViewportIndex].viewportId
+    );
+    const { element } = viewportInfo;
+    const cine = cines[activeViewportIndex];
+    if (cine.isPlaying) {
+      cineService.stopClip(element);
+      cine.isPlaying = false;
+    }
+  };
+
+  const handleSetFrameRate = newFrameRate => {
+    if (newFrameRate >= 1 && newFrameRate <= 50) {
+      setFrameRate(newFrameRate);
+    }
+    const cine = cines[0];
+    if (cine.isPlaying) {
+      handleStop();
+      handlePlay();
+    }
+  };
+
+  useEffect(() => {
+    handlePlay;
+  }, [frameRate]);
+
+  const handleLast = () => {
+    myDynamicVolume.timePointIndex = rangeValues[1] - 1;
+  };
+
+  const handleFirst = () => {
```

Comment:
you don't really need these, just apply the function on the render since you are only using them once

onChange=()=>     myDynamicVolume.timePointIndex = rangeValues[1] - 1;

Reply:
Removed handleFirst and handleLast, just applying the function on onClick